### ISSUE TITLE
Singletons not resolving dependencies correctly through child containers

### ISSUE THREAD

**Issue by saquibmian** [structuremap/structuremap#470]
I've found what might be another bug during our upgrade from SM 2 to 4 -- let me know if I'm doing something strange here. 

Let's say I have a dependency:

```csharp
public interface IDependency { }
public class RootDependency : IDependency { }
public class NestedDependency : IDependency { }
```

Then let's say I have a plugin that requires that dependency that will live as a singleton:

```csharp
public abstract class BasePlugin {
	public IDependency InjectedDependency { get; private set; }
	protected BasePlugin( IDependency d ) { InjectedDependency = d; }
}
public class SingletonPlugin : BasePlugin {
	public SingletonPlugin( IDependency d ) : base( d ) {}
}
// there are more plugins here, but I've omitted them for brevity; I can post the full source if you want it
```

If I register `SingletonPlugin` at the base along with `IDependency` at the base, but I also override `IDependency` in a nested container, then when I resolve `SingletonPlugin` from the nested container, I expect it to give me the `IDependency` registered at the base (because it has a longer lifetime than the one registered at the nested level). Here's a test that proves this *does* happen when I use `GetAllInstances<SingletonPlugin>` but *does not* happen when I call `GetInstance<SingletonPlugin>`. As far as I can tell, those two methods should behave identically. 

```csharp
[Test]
public void CheckEvenSimpler_SingletonDoesntResolveCorrectly() {
	// create nested container from root
	IContainer container;
	{
		var root = new Container();
		root.Configure( _ => {
			_.For<BasePlugin>()
				.Use<SingletonPlugin>()
				.LifecycleIs<SingletonLifecycle>();
			_.For<IDependency>()
				.Use<IDependency>( new RootDependency() );
		} );
		container = root.GetNestedContainer();
	}

	container.Configure( _ => _.For<IDependency>().Use<IDependency>( new NestedDependency() ) );

	// getting IDependency from child directly should work
	Assert.IsInstanceOf<NestedDependency>( container.GetInstance<IDependency>() );

	// getting SingletonPlugin one way -- this was done first because it behaves as expected
	foreach( var instance in container.GetAllInstances<SingletonPlugin>() ) {
		Assert.IsInstanceOf<RootDependency>( instance.InjectedDependency );
	}

	// this fails
	// getting SingletonPlugin from child should work, and get IDependency from the root
	// but this fails, because SM failed to go up to the root to resolve this, as was expected for SingletonPlugin!
	Assert.IsInstanceOf<RootDependency>( container.GetInstance<SingletonPlugin>().InjectedDependency );
}
```

**Comment by jeremydmiller** [comment]
@saquibmian,

This code is your problem:

```
Assert.IsInstanceOf<RootDependency>( container.GetInstance<SingletonPlugin>().InjectedDependency );
```

You're retrieving `SingletonPlugin` directly instead of going against the `BasePlugin` type. According to StructureMap, that's an entirely different registration. If you do a `WhatDoIHave()` on your nested container after fetching `SingletonPlugin`, you'll see that it has added a registration for that type, by that type, with the default lifecycle.

**Comment by jeremydmiller** [comment]
This as well: `container.GetAllInstances<SingletonPlugin>()`

**Comment by saquibmian** [comment]
That makes a lot of sense. Thank you!

I'm still confused about why the following is not true, though:

```csharp
container.GetInstance<SingletonPlugin>().InjectedDependency == container.GetAllInstances<SingletonPlugin>().First().InjectedDependency;
container.GetInstance<SingletonPlugin>().InjectedDependency == container.GetInstances<IEnumerable<SingletonPlugin>>().First().InjectedDependency;
```

The `EnumerableFamilyPolicy` appears to be using `AllPossibleInstance` semantics, which I'm guessing also powers `GetAllInstances` (that would explain the difference in behaviour between the two). Am I right in thinking that `AllPossibleInstances` will ask the container to build `Instance`s and downcast them, thereby never actually asking the container for the concrete type?

**Comment by jeremydmiller** [comment]
"The EnumerableFamilyPolicy appears to be using AllPossibleInstance semantics," -- no, it doesn't. It doesn't cross into other registrations and other types. It only return the instances registered for that particular element type, i.e., the "T" in IList<T> for example. StructureMap is a bit less magical in this particular regard than some of the other IoC containers.

**Comment by saquibmian** [comment]
You're right, digging into the code more, I can see it essentially forwards `IEnumerable<T>` to `GetAllInstances<T>`: https://github.com/structuremap/structuremap/blob/f23222850dc125f87a424c410e5a97f9d743a5ed/src/StructureMap.Shared/Building/AllPossibleValuesDependencySource.cs#L35

So that explains why the second conditional behaves like the first. Any ideas why the first conditional isn't true?

**Comment by jeremydmiller** [comment]
Same thing as before, you're asking for the concrete type instead of BasePlugin where the registrations are made. 

```
container.GetInstance<SingletonPlugin>().InjectedDependency == container.GetAllInstances<SingletonPlugin>().First().InjectedDependency;
```

**Comment by saquibmian** [comment]
I'm confused.

```csharp
var d1 = container.GetInstance<SingletonPlugin>().InjectedDependency; // this is NestedDependency
var d2 = container.GetAllInstances<SingletonPlugin>().First().InjectedDependency; // this is RootDependency
```

Why are `d1` and `d2` not the same? In both cases I'm querying for `SingletonPlugin`.

**Comment by jeremydmiller** [comment]
Are you trying to test reference equality between d1 and d2? It wouldn't be. Again, StructureMap is treating a request to `SingletonPlugin` as a completely different thing than if you were requesting `BasePlugin` that is marked as a singleton.

Check the Container.WhatDoIHave() after resolving `SingletonPlugin` and you'll see that there'll be an all new registration for `SingletonPlugin` with the default transient lifecycle.
